Add tidy check to deny merge commits

Noratrieb opened the pull request:
This will prevent users with the pre-push hook from pushing a merge commit.

Exceptions are added for subtree updates. These exceptions are a little hacky and may be non-exhaustive but can be extended in the future.

I added a link to @jyn514's blog post for the error case because that's the best resource to solve merge commits. But it would probably be better if it was integrated into https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy, then we could link that instead.

r? @jyn514

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3575797897/jobs/6012794647) [(plain)](https://github.com/rust-lang/rust/commit/e1dade71fa4a6a1af1c49e4f76ad4f605b64e3fb/checks/9774952902/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/tools/tidy/src/no_merge.rs at line 4:
     let Ok(remote) = get_rust_lang_rust_remote() else { return; };
     let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
-    let mut bad_merge_commits = merge_commits
-        .lines()
-        .lines()
-        .filter(|commit| {
-            !(
-                // Bors is the ruler of merge commits.
-                commit.starts_with("Auto merge of")
+    let mut bad_merge_commits = merge_commits.lines().filter(|commit| {
+        !(
+            // Bors is the ruler of merge commits.
+            commit.starts_with("Auto merge of")
                 || commit.starts_with("Rollup merge of")
                 // Our fellow subtrees are fine as well.
                 || commit.contains("Merge from rustc")
Diff in /checkout/src/tools/tidy/src/no_merge.rs at line 16:
                 || commit.contains("merge rustc history")
-                || commit.contains("rust-analyzer"))
-                || commit.contains("clippy")
-        });
+                || commit.contains("rust-analyzer")
+        ) || commit.contains("clippy")
 
 
     if let Some(merge) = bad_merge_commits.next() {
         tidy_error!(
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/tools/tidy/src/no_merge.rs" "/checkout/src/tools/tidy/src/debug_artifacts.rs" "/checkout/src/tools/tidy/src/main.rs" "/checkout/src/tools/tidy/src/unit_tests.rs" "/checkout/src/tools/tidy/src/style.rs" "/checkout/src/tools/tidy/src/errors.rs" "/checkout/src/tools/tidy/src/extdeps.rs" "/checkout/src/bootstrap/tool.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

fbstj commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,73 @@
+                // Our fellow subtrees are fine as well.
+                || commit.contains("Merge from rustc")
+                || commit.contains("merge rustc history")
+                || commit.contains("rust-analyzer"))
+                || commit.contains("clippy")
```
are these  brackets correct? are these the only tools that cause problems? should there be a note to add future tools to this?

Nilstrieb replied:
The brackets are not correct, thanks, blame rustfmt for not putting the closing one on a new line :D

And yes there should probably be a note for that

jyn514 commented:
> I added a link to @jyn514's blog post for the error case because that's the best resource to solve merge commits. But it would probably be better if it was integrated into https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy, then we could link that instead.

@Nilstrieb do you have time to add this to the dev guide?

jyn514 commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,77 @@
+            // Bors is the ruler of merge commits.
+            commit.starts_with("Auto merge of")
+                || commit.starts_with("Rollup merge of")
+                // Our fellow subtrees are fine as well.
+                // This can be extended for future subtrees.
+                || commit.contains("Merge from rustc")
+                || commit.contains("merge rustc history")
+                || commit.contains("rust-analyzer")
+                || commit.contains("clippy")
```
These concern me. "Merge from rustc" is just a convention, not generated by a tool, and if @RalfJung or @saethlin happen to use a different name for these it will block all Miri syncs unless we hardcode the new string.

Try passing this to git instead:
```
':!src/tools/rust-analyzer' ':!src/tools/clippy' ':!src/tools/rustfmt' ':!src/tools/miri'
```
It might be theoretically possible for someone to accidentally merge those and have it miss the lint, but that seems better than blocking a sync; and I think it's unlikely they'll *only* create a merge commit for the subtree and nothing else.

Actually, we can programatically generate these subtree directories, maybe we should do that instead?
```
; git log | grep git-subtree-dir | awk '{ print $2 }'
src/tools/miri
src/tools/rust-analyzer
compiler/rustc_smir
library/portable-simd
compiler/rustc_codegen_gcc
src/tools/rustfmt
compiler/rustc_codegen_cranelift
src/tools/clippy
```

jyn514 replied:
thanks for the offer :) but I don't want to impose that requirement for everyone running syncs, for it to be useful / trustworth all the other tools would need to script it as well, and I think the `:!src/tools/miri` solution is more reliable anyway.

jyn514 replied:
Future tools will need to add themselves to this check anyway, we can revisit it then. The `git-subtree-dir` commit is still going to be in the history for miri even if it switches to josh in the future.

Nilstrieb replied:
> ` git log --grep git-subtree-dir | grep git-subtree-dir | awk '{ print $2 }'`

This shell command is pretty non-portable so that's not gonna work and doing the grepping ourselves isn't gonna be amazing either. I think maintaining the list ourselves is fine.

RalfJung replied:
Yeah for Miri this works, but future tools might use josh immediately and never use `git subtree`.

jyn514 commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,77 @@
+        tidy_error!(
+            bad,
+            "found a merge commit in the history: `{merge}`.
+To resolve the issue, see the merge commit help at the bottom of this post: https://jyn.dev/2022/09/02/git-cheats.html.
```
this is kind of unwieldy but umm
```suggestion
To resolve the issue, see <https://jyn.dev/2022/09/02/git-cheats.html#i-made-a-merge-commit-but-now-someone-is-telling-me-merge-commits-arent-allowed-how-do-i-fix-this-please-i-already-spent-3-hours-on-this-i-dont-want-to-spend-3-more>
```

Nilstrieb replied:
This is beautiful

That said I will make a PR to the dev guide so we can link that instead.

Nilstrieb replied:
Update here: I discussed this with @jyn514 and we came to the conclusion that it would make more sense for them to write the dev guide help instead but that may take a while, so we shouldn't block it on this and just link the blog for now

jyn514 replied:
finally got around to writing the dev-guide side - I would prefer to wait for that to land if possible: https://github.com/rust-lang/rustc-dev-guide/pull/1536

https://github.com/rust-lang/rust/pull/105702 just landed adding your function to detect the remote, so you'll need to rebase over that anyway.

albertlarsan68 replied:
I added it as a private function in `format.rs`, maybe move it in the utilities module ?

jyn514 commented on src/tools/tidy/src/no_merge.rs:
````diff
@@ -0,0 +1,77 @@
+/// upstream        https://github.com/rust-lang/rust (push)
+/// ```
+fn get_rust_lang_rust_remote() -> Result<String, String> {
+    let mut git = Command::new("git");
+    git.args(["remote", "--verbose"]);
+    let output = git.output().map_err(|err| format!("{err:?}"))?;
+    if !output.status.success() {
+        return Err("failed to execute git remote command".to_owned());
+    }
+
+    let stdout = String::from_utf8(output.stdout).map_err(|err| format!("{err:?}"))?;
+
+    let rust_lang_remote = stdout
+        .lines()
+        .find(|remote| remote.contains("rust-lang") && remote.contains("(push)"))
+        .ok_or_else(|| "rust-lang/rust remote not found".to_owned())?;
+
+    let remote_name = rust_lang_remote
+        .split_ascii_whitespace()
+        .next()
+        .ok_or_else(|| "remote name not found".to_owned())?;
````
Don't you want the fetch URL, not the push URL? then you can use this instead:
```
; git config --local --get-regex 'remote\..*\.url'
remote.origin.url https://github.com/rust-lang/rust
remote.personal.url git@github.com:jyn514/rust.git
```
and not have to depend on exactly how `remote --verbose` formats its output.

jyn514 replied:
but see https://github.com/rust-lang/rust/pull/105058?notification_referrer_id=NT_kwDOAWiyO7M0OTU0NjkwMDUzOjIzNjM4NTg3&notifications_query=is%3Aunread#discussion_r1036235085

jyn514 commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,77 @@
+
+    let output = git.output().map_err(|err| format!("{err:?}"))?;
+    if !output.status.success() {
+        return Err("failed to execute git remote command".to_owned());
```
```suggestion
        return Err("failed to execute git log command".to_owned());
```

Nilstrieb commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,77 @@
+/// Runs `git log --merges --format=%s $REMOTE/master..HEAD` and returns all commits
+fn find_merge_commits(remote: &str) -> Result<String, String> {
+    let mut git = Command::new("git");
+    git.args(["log", "--merges", "--format=%s", &format!("{remote}/master..HEAD")]);
```
```
git rev-parse --abbrev-ref "master@{upstream}"
```
gives back `origin/master` for me, which is my outdated fork and not rust-lang/rust. Granted, this does work but it may be slower, especially if the fork master is very outdated.

Mark-Simulacrum commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,77 @@
+
+pub fn check(_: (), bad: &mut bool) {
+    let Ok(remote) = get_rust_lang_rust_remote() else { return; };
+    let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
```
I am not a fan of the silent return here. I think a better way of achieving the same aims here is to split this check out, either into a separate tool or gate it on an environment variable, and explicitly only intend that it is run *by default* in CI where we control the environment. And when run, it shouldn't ignore failures to retrieve things like the remote name etc, but rather error.

Nilstrieb replied:
The whole point behind this is to run locally before pushing anything, so it should run by default. My reasoning behind the silent return was that people with bespoke git setups are generally less likely to make such errors so the check isn't that important for them. But I agree that in CI it should error if there's anything wrong. What would the best way be to distinguish between CI and not-CI?

Mark-Simulacrum replied:
I'd be surprised if it's that common to run tidy before pushing -- it seems like a relatively unusual thing to me, but I could be surprised.

https://github.com/rust-lang/rust/blob/master/src/bootstrap/util.rs#L260

jyn514 replied:
@Mark-Simulacrum we run tidy in the pre-push hook `x.py setup ` creates

Mark-Simulacrum replied:
Hm, okay. I guess it seems like if you're running via pre-push hook, then you at least have git, right? So presumably in that case we can maybe pass something from there...

In any case, I think at minimum actually running it in CI seems like a good idea.

jyn514 reviewed (changes requested):
The git handling looks good to me (or at least, good enough it shouldn't delay this longer), but I think I agree with Mark we should run this in CI. Can you either:
- change this to fail silently locally, but give a hard error if git fails in CI (see https://github.com/rust-lang/rust/blob/dc30b92cc576ed9e097db7fece0af13662de8d8d/src/bootstrap/util.rs#L260 for how to detect it); or,
- change this to give a hard error locally if git fails

I have a slight preference for the latter but I'm ok with either.

Noratrieb committed:
Add tidy check to deny merge commits

This will prevent users with the pre-push hook from pushing a merge
commit.

Exceptions are added for subtree updates. These exceptions are a little
hacky and may be non-exhaustive but can be extended in the future.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3698054215/jobs/6263777219) [(plain)](https://github.com/rust-lang/rust/commit/0815bc93a63af6a881d694eb50efdc240b765ab0/checks/10098996362/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
* highest error code: E0791
Found 507 error codes
Found 0 error(s) in error codes
Done!
thread '<unnamed>' panicked at 'Encountered error while testing Git status: "failed to execute git log command: fatal: ambiguous argument 'origin/master..HEAD': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n"', src/tools/tidy/src/no_merge.rs:20:25
* 392 features
* 392 features
thread 'main' panicked at 'a scoped thread panicked', src/tools/tidy/src/main.rs:38:5
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3765200668/jobs/6400431421) [(plain)](https://github.com/rust-lang/rust/commit/c6d60734d4434fe5a98c91f4c15a22d3c41a81a2/checks/10272790934/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 633 error codes
* highest error code: E0791
thread '<unnamed>' panicked at 'Encountered error while testing Git status: "failed to execute git log command: fatal: ambiguous argument 'origin/master..HEAD': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n"', src/tools/tidy/src/no_merge.rs:20:25
Found 507 error codes
Found 0 error(s) in error codes
Done!
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/tools/tidy/src/main.rs:42:61
```

</details>

Noratrieb committed:
Checkout `master` branch in CI

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3766147173/jobs/6402345807) [(plain)](https://github.com/rust-lang/rust/commit/1691eee9780c32f3b7f4ed9883dc814d518e3d14/checks/10275189315/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
   Compiling ignore v0.4.18
   Compiling cargo-platform v0.1.2
   Compiling cargo_metadata v0.14.0
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
error: unused `Result` in tuple element 1 that must be used
  ::: src/tools/tidy/src/no_merge.rs:22:5
   |
   |
22 |       dbg!(output.status, String::from_utf8(output.stderr), String::from_utf8(output.stdout),);
   |
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `-D unused-must-use` implied by `-D warnings`

error: unused `Result` in tuple element 2 that must be used
  ::: src/tools/tidy/src/no_merge.rs:22:5
   |
   |
22 |       dbg!(output.status, String::from_utf8(output.stderr), String::from_utf8(output.stdout),);
   |
   |
   = note: this `Result` may be an `Err` variant, which should be handled
error: could not compile `tidy` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:00:11
```

</details>

Nilstrieb commented:
CI doesn't appear to have the `origin/master` branch locally but fetching it doesn't work because it's a read-only file system..

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3766279628/jobs/6402614044) [(plain)](https://github.com/rust-lang/rust/commit/7650f7d2cb47f91fe10c597319bf1d2ebf6a44df/checks/10275523943/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 633 error codes
* highest error code: E0791
thread '<unnamed>' panicked at 'Encountered error while testing Git status: "failed to execute git log command: error: cannot open .git/FETCH_HEAD: Read-only file system\n"', src/tools/tidy/src/no_merge.rs:20:5
Found 507 error codes
Found 0 error(s) in error codes
Done!
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/tools/tidy/src/main.rs:42:61
```

</details>

jyn514 commented:
> CI doesn't appear to have the `origin/master` branch locally but fetching it doesn't work because it's a read-only file system..

Can you fetch it using `actions/checkout`? I think you'll have to add a new `step` to the CI job because it only accepts a single `ref`, but it should be doable:
https://github.com/rust-lang/rust/blob/4566db335902599083bc622c82254f88870c6488/src/ci/github-actions/ci.yml#L101-L104
https://github.com/actions/checkout#usage

jyn514 commented:
btw, in the future please switch the labels to S-waiting-on-review so it's easier for me to find the PR.

Nilstrieb commented:
@bors rollup=iffy
this is getting very iffy

Nilstrieb commented:
i got it working with the actions checkout, thanks!
@rustbot ready

Noratrieb committed:
Add `build_helper` crate to share code between tidy and bootstrap

Nilstrieb commented:
@rustbot ready

jyn514 commented on src/tools/tidy/src/no_merge.rs:
```diff
@@ -0,0 +1,98 @@
+    Ok(stdout)
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum CiEnv {
+    /// Not a CI environment.
+    None,
+    /// The Azure Pipelines environment, for Linux (including Docker), Windows, and macOS builds.
+    AzurePipelines,
+    /// The GitHub Actions environment, for Linux (including Docker), Windows and macOS builds.
+    GitHubActions,
+}
+
+impl CiEnv {
+    /// Obtains the current CI environment.
+    pub fn current() -> CiEnv {
+        if std::env::var("TF_BUILD").map_or(false, |e| e == "True") {
+            CiEnv::AzurePipelines
+        } else if std::env::var("GITHUB_ACTIONS").map_or(false, |e| e == "true") {
+            CiEnv::GitHubActions
+        } else {
+            CiEnv::None
+        }
+    }
+
+    pub fn is_ci() -> bool {
+        Self::current() != CiEnv::None
+    }
+}
```
Can you move this to `build_helper` and change bootstrap to import it instead of duplicating it?

jyn514 reviewed (approved):
Looks great, thanks! r=me with the last nit fixed :)

Nilstrieb commented:
@rustbot ready

jyn514 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 878af66b53459659b3f53f1c26af1123e82d79a0 has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 878af66b53459659b3f53f1c26af1123e82d79a0 -->
<!-- homu: {"type":"Approved","sha":"878af66b53459659b3f53f1c26af1123e82d79a0","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 878af66b53459659b3f53f1c26af1123e82d79a0 with merge 4839886f0abe208ab8f2bb73a3076a59fe2ab60c...
<!-- homu: {"type":"BuildStarted","head_sha":"878af66b53459659b3f53f1c26af1123e82d79a0","merge_sha":"4839886f0abe208ab8f2bb73a3076a59fe2ab60c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3809466978/jobs/6481891936)
Approved by: jyn514
Pushing 4839886f0abe208ab8f2bb73a3076a59fe2ab60c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3809466978/jobs/6481891936"},"merge_sha":"4839886f0abe208ab8f2bb73a3076a59fe2ab60c"} -->

rust-timer commented:
Finished benchmarking commit (4839886f0abe208ab8f2bb73a3076a59fe2ab60c): [comparison URL](https://perf.rust-lang.org/compare.html?start=ce85c98575e3016cf2007d90a85be321e592aa96&end=4839886f0abe208ab8f2bb73a3076a59fe2ab60c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ce85c98575e3016cf2007d90a85be321e592aa96&end=4839886f0abe208ab8f2bb73a3076a59fe2ab60c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ce85c98575e3016cf2007d90a85be321e592aa96&end=4839886f0abe208ab8f2bb73a3076a59fe2ab60c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-2.1%, -2.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ce85c98575e3016cf2007d90a85be321e592aa96&end=4839886f0abe208ab8f2bb73a3076a59fe2ab60c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

klensy commented:
But this don't work? https://github.com/rust-lang/rust/pull/99339#issuecomment-1588748881
